### PR TITLE
RecycleGridLayout : Fix layout when number of widgets match number of columns

### LINKED ISSUE #7255: RecycleGridLayout not able to render viewclass when number of columns is equal to the length of data
**Software Versions**
* Python: 3.8.3
* OS: Windows 10 Home
* Kivy: 2.0.0
* Kivy installation method: installed kivy base through pip

Description:
    The recycle grid layout has an issue in rendering the viewclass objects when the number of columns are equal to the number of viewclass objects, to demonstrate this there is a demo below.

**Expected behavior**
It is supposed to display the objects of the view class in a grid, but here they disappear and only the first object is visible on-screen

**To Reproduce**
To reproduce this bug just make a recycle grid layout in a recycle view with the same number of columns as the number of the objects of the view class you want to add.

**Code and Logs and screenshots**
```python
from kivy.app import App
from kivy.lang import Builder
kv = '''
RecycleView:
    data: [{"text": "Label "+str(x)} for x in range(5)]
    viewclass: 'Button'
    RecycleGridLayout:
        cols: 5
        default_size: None, 200
        default_size_hint: 1, None
'''
class MyApp(App):
    def build(self):
        return Builder.load_string(kv)
MyApp().run()
```
Screenshot (401)

**Additional context**
This works in kivy 1.11.1, also this behavior is not present in the regular grid layout.

### PR DESCRIPTION
**Pull request by gottadiveintopython**
Maintainer merge checklist
* [x] Title is descriptive/clear for inclusion in release notes.
* [x] Applied a `Component: xxx` label.
* [ ] Applied the `api-deprecation` or `api-break` label.
* [ ] Applied the `release-highlight` label to be highlighted in release notes.
* [x] Added to the milestone version it was merged into.
* [x] **Unittests** are included in PR.
* [ ] Properly documented, including `versionadded`, `versionchanged` as needed.

fixes #7255 

I put "WIP" in the title because I feel like there is a better implemetation.

### CODE REVIEW

**Comment by tshirtman** [comment]
The parametrised test seems to repeut the implementation, i think it's better to be less smart in such case and be explicit about tho expected result, each parameter could be a tuple of (input, exptected_output) instead.

**Comment by matham** [comment]
>The parametrised test seems to repeat the implementation, i think it's better to be less smart in such cases and be explicit about tho expected result, each parameter could be a tuple of (input, exptected_output) instead.

Agree. But also, this is not testing the implementation directly, but rather an analog of the implementation. So if one changes the implementation, the test would not be helpful anymore. So it's better to test the actual implementation and use hard-coded input/output values as @tshirtman suggested.

**Comment by matham** [comment]
I think I wasn't clear enough with what I was trying to say, but could you test an actual gridlayout widget? Beause this basically tests that the current implementation is current.  But if someone changes the gridlayout in the future, the test will not test that.

**Comment by gottadiveintopython** [comment]
@tshirtman Ah sorry for my explanation being not enough. There is a situation where we cannot just sort the view indices in ascending order as the latest commit shows.

@matham No, you were clear enough. I pushed the latest commit *before* I read your comment. 

> it's better to test the actual implementation

OK, so is it ok to create a new function that implements that specific part, and use it in the unittests? or I better to use `compute_visible_views()` itself in the unittests somehow?

**Comment by gottadiveintopython** [comment]
When I implemented the 'orientation', I kept the original code as possible in order to avoid introducing new bugs. But maybe this `compute_visible_views()` should be re-written entirely (if I come up with good implementation).

**Comment by matham** [comment]
Yeah, I'd say to call `compute_visible_views` directly and ensure the it returns them in the correct order.

Or even better, create a actual recyclegridlayout (and normal gridlayout) and set the data/children, run `do_layout` or whatever is required to do the layout, and then check that the children widgets are at the expected positions. But `compute_visible_views` is enough if you don't want to do that.

**Comment by matham** [comment]
>But maybe this compute_visible_views() should be re-written entirely (if I come up with good implementation) to avoid this complicated unittests.

Ultimately, it'd be nice if we can test that the widgets are laid out correctly, irrespective of the implementation. So I'm not sure if that would help.

**Comment by matham** [comment]
Perfect. Just one more thing, it's probably better if you use the [clock fixture](https://github.com/kivy/kivy/blob/master/kivy/tests/fixtures.py#L11) rather than the raw clock.

**Comment by gottadiveintopython** [comment]
To use that fixture I have to do something like this:

```python
class Test_layout:
    def test_case1(self, kivy_clock, ...):
       self.compute_layout(..., clock=kivy_clock)
    def test_case2(self, kivy_clock, ...):
       self.compute_layout(..., clock=kivy_clock)
```

This isn't desireble, right?

**Comment by matham** [comment]
>This is not desirable, isn't it?

I think that's fine!? It's just more verbose, but I don't really see another way that is not hacky.

**Comment by gottadiveintopython** [comment]
@matham OK, how about the following fixture? Do you think it's a good idea?

```python
# kivy/tests/fixtures.py
@pytest.fixture(scope='module')
def module_scope_kivy_clock():
    # same as kivy_clock


# kivy/tests/test_uix_recyclegridlayout.py
global_kivy_clock = None

@pytest.fixture(scope='module', autouse=True)
def acquire_kivy_clock(module_scope_kivy_clock):
    global global_kivy_clock
    global_kivy_clock = module_scope_kivy_clock
    yield module_scope_kivy_clock
    global_kivy_clock = None

class Test_layout:
    def compute_layout(...):
         global_kivy_clock.tick()
```

**Comment by gottadiveintopython** [comment]
`TestLayout_all_the_data_is_visible` might be unnecessary if `TestLayout_only_a_part_of_the_data_is_visible` exists.

**Comment by matham** [comment]
>OK, how about the following fixture?

I think the way it is currently is better. It's simpler even if more verbose, which is not really a big deal.

>TestLayout_all_the_data_is_visible might be unnecessary if TestLayout_only_a_part_of_the_data_is_visible exists.

Extra tests have never hurt anyone 😄

**Review by matham — DISMISSED** [review]

**Comment by matham** [comment]
Feel free to merge if you're happy with it.

**Comment by gottadiveintopython** [comment]
@matham I think the pr is mergable now. I'd appreciate if you approve it again :)

File: kivy/uix/recyclegridlayout.py [hunk 1]
```diff
@@ -220,22 +222,24 @@ def compute_visible_views(self, data, viewport):
...
             at_idx((x, y)),
             at_idx((right, y)),
             at_idx((x, top)),
             at_idx((right, top)),
         ))
 
         n = len(data)
+        if len({tl, tr, bl, br}) < 4:
+            # visible area is one row/column
+            return list(range(min(n, tl), min(n, br + 1)))
```
Comment (gottadiveintopython) [inline kivy/uix/recyclegridlayout.py:235]: This part doesn't have to create a list. I have to fix it.

Comment (matham) [reply kivy/uix/recyclegridlayout.py]: Hmm I suppose. The interface doesn't say that it needs to return a list so it should be fine.

**Review by matham — APPROVED** [review]

**Pull request merged**